### PR TITLE
SDL 0274: Add preferred FPS to VideoStreamingCapability

### LINKED ISSUE #115: [SDL 0274] Add preferred FPS to VideoStreamingCapability
Proposal: [Add preferred FPS to VideoStreamingCapability](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0274-add-preferred-FPS.md)

> This proposal extends SDL video streaming feature by adding preferred FPS to `VideoStreamingCapability`, so that frame rate of video streaming can be adjusted per OEM.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/913

Steering Committee Decision:

> The Steering Committee voted to accept this proposal with revisions.  The revisions will include the items listed in [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/913#issuecomment-579062117) from the author.

> The proposal .md file was updated to reflect these revisions on 2/10/2020.

### PR DESCRIPTION
**Pull request by shiniwat**
Fixes #115

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
VideoStreamingParametersTest will partially cover this change.

### Summary
Implementing SDL 0274: Add preferred FPS to VideoStreamingCapability.

##### Breaking Changes
* N/A

##### Enhancements
* Adding preferredFPS into VideoStreamingCapability.

##### Bug Fixes
* N/A

### Tasks Remaining:
- N/A

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform) and this pull request adheres to the [Contributing Guide](./CONTRIBUTING.md).

### CODE REVIEW

**Comment by jordynmackool** [comment]
Hi @shiniwat, since the core PR for this feature is not complete, it seems that this PR would have not been tested fully. To ensure the review process goes smoothly please provide the Core branch/PR that this was tested against.

Once this PR is fully tested, tag me and Livio will review. Thank you!

**Comment by shiniwat** [comment]
Hi @jordynmackool,
The core PR is still in review at https://github.com/smartdevicelink/sdl_core/pull/3437, but I've been running my locally-built core for a while. Regarding this PR, as we discussed in sdl_evolution review, I believe sdl_javascript_suite still does not support VPM. All I can do with this PR is the newly added RPC parameter is handled in sdl_javascript_suite platform without errors.

**Comment by jordynmackool** [comment]
Hi @shiniwat, thank you for the reminder about the VPM. For this branch, confirming that the newly added RPC parameter is handled in the sdl_javascript_suite platform without errors would be sufficient testing. 

If you can verify that has been done, we will begin our review as soon as we can.

**Comment by shiniwat** [comment]
Hi @jordynmackool, this PR is ready for review by running with sdl_core PR at https://github.com/smartdevicelink/sdl_core/pull/3437.

File: .gitmodules [hunk 1]
```diff
@@ -1,4 +1,4 @@
 [submodule "lib/rpc_spec"]
```
Comment (crokita) [inline .gitmodules:1]: This comment is a reminder to undo this change once the rpc_spec includes the new feature in develop

File: lib/js/src/streaming/video/_VideoStreamingParameters.js [hunk 2]
```diff
@@ -104,6 +104,10 @@ class _VideoStreamingParameters {
         if (formats !== null && formats.length > 0) {
             this._format = formats[0];
         }
+        const fps = capability.getPreferredFPS();
+        if (fps != null) {
```
Comment (crokita) [inline lib/js/src/streaming/video/_VideoStreamingParameters.js:108]:
```suggestion
        if (fps !== null) {
```
Linter rule requires use of `!==`

File: tests/node/streaming/video/VideoStreamingParametersTest.js [hunk 3]
```diff
@@ -91,9 +91,10 @@ describe('VideoStreamingParametersTest', function () {
 
         format = new VideoStreamingFormat(VideoStreamingProtocol.RTP, VideoStreamingCodec.H264);
         capability.setSupportedFormats([format]);
+        capability.setPreferredFPS(15);
         params.update(capability);
         Validator.assertEquals(params.getFormat(), format);
```
Comment (crokita) [inline tests/node/streaming/video/VideoStreamingParametersTest.js:97]: `setPreferredFPS` method added to the test but there's no test to check it got set with the update.
Needs `Validator.assertEquals(params.getFrameRate(), 15);`

**Review by crokita — COMMENTED** [review]
See feedback @shiniwat

File: .gitmodules (hunk 1 above)
Comment (shiniwat) [reply .gitmodules:1]: Updated rpc_spec commit hash, and restored the .gitmodules accordingly.

File: lib/js/src/streaming/video/_VideoStreamingParameters.js (hunk 2 above)
Comment (shiniwat) [reply lib/js/src/streaming/video/_VideoStreamingParameters.js]: Sorry for the typo. Updated.

File: tests/node/streaming/video/VideoStreamingParametersTest.js (hunk 3 above)
Comment (shiniwat) [reply tests/node/streaming/video/VideoStreamingParametersTest.js:98]: That makes sense. Updated accordingly,

File: .gitmodules [hunk 4]
```diff
@@ -1,4 +1,3 @@
...
 	path = lib/rpc_spec
 	url = https://github.com/smartdevicelink/rpc_spec.git
-	branch = master
```
Comment (crokita) [inline .gitmodules:4]: Please keep the branch property in this file. This file should be unchanged in the end.

**Review by crokita — COMMENTED** [review]
See additional feedback @shiniwat

File: .gitmodules (hunk 4 above)
Comment (shiniwat) [reply .gitmodules]: Well, I have restored the .gitmodules. Thanks.

**Review by crokita — APPROVED** [review]
Looks good. Will wait on https://github.com/smartdevicelink/rpc_spec/pull/229 to be merged before continuing
